src/cpp/fastdds/builtin/type_lookup_service/TypeLookupManager.cpp

```diff
@@ -235,50 +274,167 @@ void TypeLookupManager::remove_remote_endpoints(
+        id = request->header().requestId();
     }
     type.deleteData(request);
     return id;
 }
 
 ReturnCode_t TypeLookupManager::async_get_type(
-        xtypes::TypeInformation,
-        GuidPrefix_t,
-        AsyncGetTypeCallback&)
+        const xtypes::TypeInformation& type_information,
+        const fastrtps::rtps::GuidPrefix_t& type_server,
+        const AsyncGetTypeCallback& callback)
+{
+    ReturnCode_t solve_ret = check_type_identifier_received(
+        type_information.complete().typeid_with_size(), type_server);
+    if (RETCODE_OK == solve_ret)
+    {
+        // The type is already known, invoke the callback
+        callback();
+    }
+    else if (RETCODE_NO_DATA == solve_ret)
+    {
+        // The type needs to be queried, add callback to notify once the remote type has been discovered.
+        add_async_get_type_callback(type_information.complete().typeid_with_size(), type_server, callback);
+    }
+
+    return solve_ret;
+}
+
+ReturnCode_t TypeLookupManager::check_type_identifier_received(
+        const xtypes::TypeIdentfierWithSize& type_identifier_with_size,
+        const fastrtps::rtps::GuidPrefix_t& type_server)
+{
+    // Check if the type is known
+    if (fastrtps::rtps::RTPSDomainImpl::get_instance()->type_object_registry_observer().
+                    is_type_identifier_known(type_identifier_with_size))
+    {
+        return RETCODE_OK;
+    }
+
+    // Check if the TypeInformation was received before
+    if (async_get_type_callbacks_.find(type_identifier_with_size) != async_get_type_callbacks_.end())
+    {
+        // Return without sending new request
+        return RETCODE_NO_DATA;
+    }
+
+    // TypeInformation is unknown
+    xtypes::TypeIdentifierSeq unknown_type;
+    unknown_type.push_back(type_identifier_with_size.type_id());
+    SampleIdentity get_type_dependencies_request = get_type_dependencies(unknown_type, type_server);
+    if (INVALID_SAMPLE_IDENTITY != get_type_dependencies_request)
+    {
+        // Store the sent requests and associated TypeIdentfierWithSize
+        add_async_get_type_request(get_type_dependencies_request, type_identifier_with_size);
+        return RETCODE_NO_DATA;
+    }
+    else
+    {
+        // Failed to send request, return error
+        return RETCODE_ERROR;
+    }
+}
+
+bool TypeLookupManager::add_async_get_type_callback(
+        const xtypes::TypeIdentfierWithSize& type_identifier_with_size,
+        const fastrtps::rtps::GuidPrefix_t& type_server,
+        const AsyncGetTypeCallback& callback)
+{
+    (void)type_server;
+    std::unique_lock<std::mutex> lock(async_get_types_mutex_);
+    try
+    {
+        // Check if TypeIdentfierWithSize already exists in the map
+        auto it = async_get_type_callbacks_.find(type_identifier_with_size);
+        if (it != async_get_type_callbacks_.end())
+        {
+            // TypeIdentfierWithSize exists, add the callback
+            it->second.push_back(callback);
+        }
```

Comment:
This search has already been performed in `check_type_identifier_received`. I suggest  adding the callback in that other method so the same search is avoided and done only once.

Reply:
Removed method in [cfa3f78](https://github.com/eProsima/Fast-DDS/pull/4177/commits/cfa3f781c09d678528b673acc9feef170c546bc0)